File: context_os/config/config_manager.py
"""ConfigManager — 支持 mtime 热加载。"""
from __future__ import annotations
import logging, os, time, re, yaml
from pathlib import Path
from threading import Lock, Thread
from typing import Optional
from context_os.config.app_config import AppConfig, from_dict

logger = logging.getLogger(__name__)

class ConfigManager:
    def __init__(self, config_path: str = "config.yaml"):
        self._path = Path(config_path)
        self._lock = Lock()
        self._current: Optional[AppConfig] = None
        self._last_modified: float = 0
        self._watcher_thread: Optional[Thread] = None
        self._running = False
        self._reload()

    def get(self) -> AppConfig:
        if self._current is None: self._reload()
        return self._current

    def check_for_update(self) -> bool:
        if not self._path.exists(): return False
        lm = self._path.stat().st_mtime
        if lm > self._last_modified:
            self._reload()
            return True
        return False

    def start_watching(self, interval_sec: int = 30) -> None:
        if self._running: return
        self._running = True
        self._watcher_thread = Thread(target=self._watch_loop, args=(interval_sec,), daemon=True)
        self._watcher_thread.start()
        logger.info("ConfigWatcher started: %s every %ds", self._path, interval_sec)

    def stop_watching(self) -> None:
        self._running = False

    def _reload(self) -> None:
        with self._lock:
            if not self._path.exists():
                logger.warning("Config not found: %s, using defaults", self._path)
                self._current = AppConfig()
                self._current.loaded_at = time.time()
                return
            try:
                raw = self._path.read_text(encoding="utf-8")
                resolved = self._resolve_env_vars(raw)
                data = yaml.safe_load(resolved) or {}
                cfg = from_dict(data)
                cfg.loaded_at = time.time()
                self._current = cfg
                self._last_modified = self._path.stat().st_mtime
                logger.info("Config loaded: %s (%d bytes)", self._path, len(raw))
            except Exception as e:
                logger.error("Failed to load config: %s", e)

    def _watch_loop(self, interval_sec: int) -> None:
        while self._running:
            try:
                if self.check_for_update():
                    logger.info("Config hot-reloaded: %s", self._path)
            except Exception as e:
                logger.warning("Config watch error: %s", e)
            time.sleep(interval_sec)

    @staticmethod
    def _resolve_env_vars(value: str) -> str:
        _pat = re.compile(r"\$\{([^}:]+)(?::([^}]*))?\}")
        def _rep(m: re.Match) -> str:
            return os.environ.get(m.group(1), m.group(2) or "")
        return _pat.sub(_rep, value)

File: java-context-os/src/main/java/com/owencli/contextos/core/config/ConfigManager.java
package main.java.com.owencli.contextos.core.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.owencli.contextos.core.config.model.AppConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * ConfigManager — 分层配置管理器，支持 mtime 热加载。
 * <p>
 * 参考 DeerFlow AppConfig.from_file() + mtime 检测设计。
 * <p>
 * 使用方式:
 * <pre>
 * var cm = new ConfigManager("config.yaml");
 * cm.startWatching();         // 每 30s 检测 mtime
 * AppConfig cfg = cm.get();   // 总是返回最新配置
 * cm.stopWatching();
 * </pre>
 */
public class ConfigManager implements AutoCloseable {

    private static final Logger log = LoggerFactory.getLogger(ConfigManager.class);
    private static final Pattern ENV_VAR_PATTERN = Pattern.compile("\\$\\{([^}:]+)(?::([^}]*))?\\}");

    private final ObjectMapper mapper;
    private final Path configPath;
    private final ScheduledExecutorService scheduler;

    private volatile AppConfig currentConfig;
    private volatile long lastModified;

    public ConfigManager(String configPath) {
        this.mapper = new ObjectMapper(new YAMLFactory());
        this.configPath = Paths.get(configPath != null ? configPath : "config.yaml");
        this.scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "config-watcher");
            t.setDaemon(true);
            return t;
        });
        reload();
    }

    /** 获取当前配置。 */
    public AppConfig get() {
        return currentConfig;
    }

    /** 检查配置文件是否有更新。 */
    public boolean checkForUpdate() {
        File file = configPath.toFile();
        if (!file.exists()) return false;
        long lm = file.lastModified();
        if (lm > lastModified) {
            reload();
            return true;
        }
        return false;
    }

    /** 开始定时检测（每 30s）。 */
    public void startWatching() {
        scheduler.scheduleAtFixedRate(() -> {
            try {
                if (checkForUpdate()) {
                    log.info("Config hot-reloaded: {}", configPath);
                }
            } catch (Exception e) {
                log.warn("Config watch error: {}", e.getMessage());
            }
        }, 30, 30, TimeUnit.SECONDS);
        log.info("ConfigWatcher started: watching {} every 30s", configPath);
    }

    /** 停止检测。 */
    public void stopWatching() {
        scheduler.shutdown();
    }

    /** 重新加载配置。 */
    public synchronized void reload() {
        File file = configPath.toFile();
        if (!file.exists()) {
            log.warn("Config file not found: {}, using defaults", configPath);
            currentConfig = new AppConfig();
            currentConfig.setLoadedAt(System.currentTimeMillis());
            return;
        }
        try {
            String raw = Files.readString(configPath, StandardCharsets.UTF_8);
            String resolved = resolveEnvVars(raw);
            AppConfig cfg = mapper.readValue(resolved, AppConfig.class);
            cfg.setLoadedAt(System.currentTimeMillis());
            this.currentConfig = cfg;
            this.lastModified = file.lastModified();
            log.info("Config loaded: {} ({} bytes)", configPath, raw.length());
        } catch (Exception e) {
            log.error("Failed to load config: {}", e.getMessage(), e);
        }
    }

    /** 替换 ${VAR:default} 环境变量。 */
    static String resolveEnvVars(String input) {
        Matcher m = ENV_VAR_PATTERN.matcher(input);
        StringBuilder sb = new StringBuilder();
        while (m.find()) {
            String var = m.group(1);
            String def = m.group(2);
            String val = System.getenv(var);
            if (val == null) val = System.getProperty(var);
            if (val == null) val = def;
            m.appendReplacement(sb, Matcher.quoteReplacement(val != null ? val : ""));
        }
        m.appendTail(sb);
        return sb.toString();
    }

    @Override
    public void close() {
        stopWatching();
    }
}
